packages/core/src/storage/IndyStorageService.ts

```diff
@@ -88,115 +91,147 @@ export class IndyStorageService<T extends BaseRecord> implements StorageService<
           recordType: record.type,
           cause: error,
         })
-      } // Other indy error
-      else if (isIndyError(error)) {
-        handleIndyError(error)
       }
 
-      throw error
+      throw isIndyError(error) ? new IndySdkError(error) : error
     }
   }
 
-  /** @inheritDoc {StorageService#getById} */
+  /** @inheritDoc */
   public async getById(recordClass: BaseRecordConstructor<T>, id: string): Promise<T> {
     try {
-      const record = await this.wallet.getWalletRecord(recordClass.type, id, IndyStorageService.DEFAULT_QUERY_OPTIONS)
+      const record = await this.indy.getWalletRecord(
+        this.wallet.walletHandle,
+        recordClass.type,
+        id,
+        IndyStorageService.DEFAULT_QUERY_OPTIONS
+      )
       return this.recordToInstance(record, recordClass)
     } catch (error) {
       if (isIndyError(error, 'WalletItemNotFound')) {
         throw new RecordNotFoundError(`record with id ${id} not found.`, {
           recordType: recordClass.type,
           cause: error,
         })
-      } else if (isIndyError(error)) {
-        handleIndyError(error)
       }
 
-      throw error
+      throw isIndyError(error) ? new IndySdkError(error) : error
     }
   }
 
-  /** @inheritDoc {StorageService#getAll} */
+  /** @inheritDoc */
   public async getAll(recordClass: BaseRecordConstructor<T>): Promise<T[]> {
-    const recordIterator = await this.wallet.search(recordClass.type, {}, IndyStorageService.DEFAULT_QUERY_OPTIONS)
+    const recordIterator = this.search(recordClass.type, {}, IndyStorageService.DEFAULT_QUERY_OPTIONS)
     const records = []
     for await (const record of recordIterator) {
       records.push(this.recordToInstance(record, recordClass))
     }
     return records
   }
 
-  /** @inheritDoc {StorageService#findByQuery} */
+  /** @inheritDoc */
   public async findByQuery(
     recordClass: BaseRecordConstructor<T>,
     query: Partial<ReturnType<T['getTags']>>
   ): Promise<T[]> {
     const indyQuery = this.transformFromRecordTagValues(query as unknown as TagsBase)
 
-    const recordIterator = await this.wallet.search(
-      recordClass.type,
-      indyQuery,
-      IndyStorageService.DEFAULT_QUERY_OPTIONS
-    )
+    const recordIterator = await this.search(recordClass.type, indyQuery, IndyStorageService.DEFAULT_QUERY_OPTIONS)
     const records = []
     for await (const record of recordIterator) {
       records.push(this.recordToInstance(record, recordClass))
     }
     return records
   }
+
+  private async *search(
+    type: string,
+    query: WalletQuery,
+    { limit = Infinity, ...options }: WalletSearchOptions & { limit?: number }
+  ) {
+    try {
+      const searchHandle = await this.indy.openWalletSearch(this.wallet.walletHandle, type, query, options)
+
+      let records: Indy.WalletRecord[] = []
+
+      // Allow max of 256 per fetch operation
+      const chunk = limit ? Math.min(256, limit) : 256
```

Comment:
This chunk finder seems to be reused somewhere else. Might it be possible to extract it? If the max changes it could get weird.

Reply:
I've thought about it, however implementations differ a bit so choose to not do it for now